miasm2/ir/ssa.py

```diff
@@ -339,3 +339,267 @@ def transform(self, path):
+            self._rename_phi_lhs(block)
+
+            '''Transform all non-phi expressions into SSA'''
+            self._rename_expressions(block)
+
+            '''Update the successor's phi functions' RHS with current SSA variables'''
+            # walk over block's successors in the CFG
+            for successor in self.graph.successors_iter(block):
+                self._rename_phi_rhs(successor)
+
+            '''Save current SSA variable stack for successors in the dominator tree'''
+            for successor in dominator_tree.successors_iter(block):
+                stack.append(self._stack_rhs.copy())
+
+    def _rename_phi_lhs(self, block):
+        """
+        Transforms phi function's expressions of an IRA block
+        on the left hand side into SSA
+        :param block: IRA block label
+        """
+        if block in self._phinodes:
+            # create temporary list of phi function assignments for inplace renaming
+            tmp = list(self._phinodes[block])
+
+            # iterate over all block's phi nodes
+            for dst in tmp:
+                # transform variables on LHS inplace
+                self._phinodes[block][self._transform_expression_lhs(dst)] = self._phinodes[block].pop(dst)
+
+    def _rename_phi_rhs(self, successor):
+        """
+        Transforms the right hand side of each successor's phi function
+        into SSA. Each transformed expression of a phi function's
+        right hand side is of the form
+
+        phi(<var>.<index 1>, <var>.<index 2>, ..., <var>.<index n>)
+
+        :param successor: label of block's direct successor in the CFG
+        """
+        # if successor is in block's dominance frontier
+        if successor in self._phinodes:
+            # walk over all variables on LHS
+            for dst in self._phinodes[successor]:
+                # transform RHS expression into SSA
+                src = self._phinodes[successor][dst]
+                src_ssa = self._transform_phi_rhs(dst)
+
+                # phi function is empty
+                if isinstance(src, m2_expr.ExprId) and src.name == "phi":
+                    e = self._fill_phi(src_ssa)
+                # phi function contains at least one value
+                else:
+                    e = self._fill_phi(src_ssa, *src.args)
+
+                # update phi function
+                self._phinodes[successor][dst] = e
+
+    def _insert_phi(self):
+        """Inserts phi functions into the list of SSA expressions"""
+        for block in self._phinodes:
+            for dst in self._phinodes[block]:
+                self.expressions[dst] = self._phinodes[block][dst]
+
+    def _convert_phi(self):
+        """Inserts corresponding phi functions inplace
+        into IRA block at the beginning"""
+        for block in self._phinodes:
+            ib = self.get_block(block)
+            # list of instructions
+            instructions = []
```

Comment:
Instead of using an intermediate `instructions` object, build a group of `ExprAff` and finally build an `AssignBlock` from it, you can directly declare an `AssignBlock`, and then assign values to it like a dictionnary:
```Python
assignblk = AssignBlock()
for dst in self._phinodes[block]:
    ...
    assignblk[dst] = src
```